
<h1 align="center"> Battle of the Toronto Neighbourhoods </h1>

This notebook contains all three parts of the assignment. 


# Part 1

## Importing Data

In [66]:
import requests
import pandas as pd

In [67]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [68]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

In [69]:
len(wiki_data), type(wiki_data)

(3, list)

We only need the first table. Dropping the other tables.

In [70]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                         Not assigned  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
10                                        Not assigned  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                        Not assigned  
16                                        Not assigned  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                        Not assigned  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                        Not assigned  
25                                        Not assigned  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                        Not assigned  

Dropping Boroughs which are not assigned

In [71]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]

Grouping the records based on Postal Code

In [72]:
df = df.groupby(['Postal Code']).head()
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Checking for number of records where Neighbourhood is "Not assigned"

In [73]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [74]:
df = df.reset_index()
df.head()

index Postal Code           Borough  \
0      2         M3A        North York   
1      3         M4A        North York   
2      4         M5A  Downtown Toronto   
3      5         M6A        North York   
4      6         M7A  Downtown Toronto   

                                 Neighbourhood  
0                                    Parkwoods  
1                             Victoria Village  
2                    Regent Park, Harbourfront  
3             Lawrence Manor, Lawrence Heights  
4  Queen's Park, Ontario Provincial Government

In [81]:
df.drop(['index'], axis = 'columns', inplace = True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [82]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [15]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder # import geocoder

Tried the below approach below but the code did not complete.  Using csv file instead.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Importing the CSV file from the URL

In [76]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [84]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can join on the postal codes to get the required data.

In [83]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [79]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [85]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [86]:
combined_data.shape

(103, 5)

**Solution:** As expected, there are 103 rows from the inner join.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [22]:
!pip install geopy

In [35]:
from geopy.geocoders import Nominatim 

In [87]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [88]:
import folium

In [89]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [90]:
CLIENT_ID = 'EDS1WXC30RRBFQPMHY215FNCZ04EAEKKS5KMLNZICBF0F1IY' # your Foursquare ID
CLIENT_SECRET = 'PPZCQD24CKM5CBLWQE3ZZ25ZA4Z333HFJNUI21NL5FMM13XW' # your Foursquare Secret
ACCESS_TOKEN = 'O1DPXUYLG0W44VKNBJODKTWAD20VLADNF5ONCFEECNAGZYJY' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: EDS1WXC30RRBFQPMHY215FNCZ04EAEKKS5KMLNZICBF0F1IY
CLIENT_SECRET:PPZCQD24CKM5CBLWQE3ZZ25ZA4Z333HFJNUI21NL5FMM13XW


Next, we create a function to get all the venue categories in Toronto

In [91]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [92]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [93]:
venues_in_toronto.shape

(1322, 5)

So we have 1322 records and 5 columns. Checking sample data


In [94]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [95]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
5                                      Victoria Village   
6                                      Victoria Village   
8                             Regent Park, Harbourfront   
9                             Regent Park, Harbourfront   
10                            Regent Park, Harbourfront   
11                            Regent Park, Harbourfront   
12                            Regent Park, Harbourfront   
38                     Lawrence Manor, Lawrence Heights   
39                     Lawrence Manor, Lawrence Heights   
40                     Lawrence Manor, Lawrence Heights   
41                     Lawrence Manor, Lawrence Heights   
42                     Lawrence Manor, Lawrence Heights   
51          Queen's Park, Ontario Provincial Government   
52          Queen's Park, Ontario Provincial Government   
53          Queen's Park, Ontario Provincial Government   
54          Queen's Park, Ontario Provincial Government   
55          Queen's Park, Ontario Provincial Government   
81                                       Malvern, Rouge   
82                                            Don Mills   
83                                            Don Mills   
84                                            Don Mills   
85                                            Don Mills   
86                      Parkview Hill, Woodbine Gardens   
87                      Parkview Hill, Woodbine Gardens   
88                      Parkview Hill, Woodbine Gardens   
...                                                 ...   
1195                Northwest, West Humber - Clairville   
1196                        St. James Town, Cabbagetown   
1197                        St. James Town, Cabbagetown   
1198                        St. James Town, Cabbagetown   
1199                        St. James Town, Cabbagetown   
1200                        St. James Town, Cabbagetown   
1226             First Canadian Place, Underground city   
1227             First Canadian Place, Underground city   
1228             First Canadian Place, Underground city   
1229             First Canadian Place, Underground city   
1230             First Canadian Place, Underground city   
1256      The Kingsway, Montgomery Road, Old Mill North   
1257      The Kingsway, Montgomery Road, Old Mill North   
1258                               Church and Wellesley   
1259                               Church and Wellesley   
1260                               Church and Wellesley   
1261                               Church and Wellesley   
1262                               Church and Wellesley   
1288  Business reply mail Processing Centre, South C...   
1289  Business reply mail Processing Centre, South C...   
1290  Business reply mail Processing Centre, South C...   
1291  Business reply mail Processing Centre, South C...   
1292  Business reply mail Processing Centre, South C...   
1304  Old Mill South, King's Mill Park, Sunnylea, Hu...   
1305  Old Mill South, King's Mill Park, Sunnylea, Hu...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   
1310  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.725882               -79.315572   
3                  43.725882               -79.315572   
4                  43.725882               -79.315572   
5                  43.725882             

So there are 414 records for each neighbourhood.

Checking for the maximum venue categories

In [96]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                                     
Accessories Store                                              Wexford, Maryvale   
Airport                                                                Downsview   
Airport Food Court             CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate                   CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge                 CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service                CN Tower, King and Spadina, Railway Lands, Har...   
Airport Terminal               CN Tower, King and Spadina, Railway Lands, Har...   
American Restaurant                     Toronto Dominion Centre, Design Exchange   
Antique Shop                                       High Park, The Junction South   
Aquarium                       Harbourfront East, Union Station, Toronto Islands   
Art Gallery                             Toronto Dominion Centre, Design Exchange   
Art Museum                                                    Central Bay Street   
Arts & Crafts Store                                  Willowdale, Willowdale East   
Asian Restaurant                                        Richmond, Adelaide, King   
Athletics & Sports                                              Woodbine Heights   
Auto Garage                                                    Wexford, Maryvale   
Auto Workshop                  Business reply mail Processing Centre, South C...   
BBQ Joint                                    The Annex, North Midtown, Yorkville   
Baby Store                                                              Christie   
Bagel Shop                     Summerhill West, Rathnelly, South Hill, Forest...   
Bakery                                                         Wexford, Maryvale   
Bank                                              University of Toronto, Harbord   
Bar                                               University of Toronto, Harbord   
Baseball Field                 Old Mill South, King's Mill Park, Sunnylea, Hu...   
Basketball Stadium             Harbourfront East, Union Station, Toronto Islands   
Beer Bar                                          University of Toronto, Harbord   
Beer Store                                                      Woodbine Heights   
Belgian Restaurant                     Kensington Market, Chinatown, Grange Park   
Bike Shop                                                                Leaside   
Bistro                                                               Berczy Park   
...                                                                          ...   
Sporting Goods Shop                            North Toronto West, Lawrence Park   
Sports Bar                                                               Leaside   
Stadium                             Brockton, Parkdale Village, Exhibition Place   
Stationery Store                                                 Studio District   
Steakhouse                                           Willowdale, Willowdale East   
Supermarket                                                     Thorncliffe Park   
Supplement Shop                Mimico NW, The Queensway West, South of Bloor,...   
Sushi Restaurant                                     Willowdale, Willowdale East   
Swim School                                                        Lawrence Park   
Tailor Shop                                                       Stn A PO Boxes   
Taiwanese Restaurant                                 St. James Town, Cabbagetown   
Tanning Salon                  Mimico NW, The Queensway West, South of Bloor,...   
Tea Room                                  First Canadian Place, Underground city   
Tennis Court                                                  Humewood-Cedarvale   
Thai Restaurant                                                  Studio District   
Theater                            

There are around 237 different types of Venue Categories. 

## One Hot encoding the venue Categories

In [97]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat.head()

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Train Station  Turkish Restaurant  \
0             0         0  ...              0                   0   
1             0         0  ...              0                   0   
2             0         0  ...              0                   0   
3             0         0  ...              0                   0   
4             0         0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 237 columns]

Adding the neighbourhood to the encoded dataframe

In [98]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [99]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories

In [100]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [101]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [102]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1             Sandwich Place              Pharmacy           Coffee Shop   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store          Dessert Shop     Electronics Store   
1                   Pub                   Gym       Airport Service   
2         Shopping Mall                 Diner        Sandwich Place   
3           Yoga Studio          Dessert Shop     Electronics Store   
4             Juice Bar               Butcher            Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant          Dumpling Restaurant   
1                  Escape Room  Eastern European Restaurant   
2                Deli / Bodega                   Restaurant   
3  Eastern European Restaurant          Dumpling Restaurant   
4                         Café            Indian Restaurant   

  10th Most Common Venue  
0              Drugstore  
1    Dumpling Restaurant  
2            Supermarket  
3              Drugstore  
4       Sushi Restaurant

Let's make the model to cluster our Neighbourhoods

In [103]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [104]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [105]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 4, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1,
       0, 1, 1, 1, 1, 1, 0, 2])

Let's add the clustering Label column to the top 10 common venue categories

In [106]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [107]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Department Store   
1           Coffee Shop          Intersection          Hockey Arena   
2                  Park                Bakery                   Pub   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio                 Diner        Sandwich Place   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0       Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1   Portuguese Restaurant            French Restaurant            Comic Shop   
2          Breakfast Spot               Chocolate Shop   Distribution Center   
3  Furniture / Home Store                  Event Space           Coffee Shop   
4   Portuguese Restaurant                         Park    Mexican Restaurant   

     8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                Drugstore             Donut Shop                Dog Run  
1  Comfort Food Restaurant    Dumpling Restaurant              Drugstore  
2                      Spa           Dessert Shop            Event Space  
3            Women's Store  Vietnamese Restaurant       Airport Terminal  
4       Italian Restaurant             Hobby Shop    Fried Chicken Joint

Drop all the NaN values to prevent data skew

In [108]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [109]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [110]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [111]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio      Department Store   
21               Women's Store                  Pool           Yoga Studio   
35           Convenience Store          Intersection           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
61                 Swim School              Bus Line           Yoga Studio   
64                 Yoga Studio         Deli / Bodega     Electronics Store   
66           Convenience Store           Yoga Studio      Department Store   
85                        Park                Bakery          Intersection   
91                       Trail            Playground           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21                Deli / Bodega  Eastern European Restaurant   
35             Department Store            Electronics Store   
49             Department Store            Electronics Store   
61             Department Store  Eastern European Restaurant   
64  Eastern European Restaurant          Dumpling Restaurant   
66            Electronics Store  Eastern European Restaurant   
85                  Yoga Studio             Department Store   
91             Department Store                 Dessert Shop   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Dumpling Restaurant             Drugstore            Donut Shop   
21          Dumpling Restaurant             Drugstore            Donut Shop   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
49  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
61          Dumpling Restaurant             Drugstore            Donut Shop   
64                    Drugstore            Donut Shop               Dog Run   
66          Dumpling Restaurant             Drugstore            Donut Shop   
85  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
91           Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common Venue  
0                 Dog Run  
21                Dog Run  
35             Donut Shop  
49             Donut Shop  
61                Dog Run  
64    Distribution Center  
66                Dog Run  
85             Donut Shop  
91                Dog Run

Cluster 2

In [112]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York             1.0                 Pizza Place   
2    Downtown Toronto             1.0                 Coffee Shop   
3          North York             1.0              Clothing Store   
4    Downtown Toronto             1.0                 Coffee Shop   
7          North York             1.0                         Gym   
8           East York             1.0                 Pizza Place   
9    Downtown Toronto             1.0                        Café   
10         North York             1.0                 Pizza Place   
12        Scarborough             1.0  Construction & Landscaping   
13         North York             1.0                         Gym   
14          East York             1.0                Skating Rink   
15   Downtown Toronto             1.0                   Gastropub   
16               York             1.0                Tennis Court   
17          Etobicoke             1.0                 Pizza Place   
18        Scarborough             1.0              Medical Center   
19       East Toronto             1.0                Neighborhood   
20   Downtown Toronto             1.0                    Beer Bar   
22        Scarborough             1.0                 Coffee Shop   
23          East York             1.0                 Coffee Shop   
24   Downtown Toronto             1.0                 Coffee Shop   
25   Downtown Toronto             1.0               Grocery Store   
26        Scarborough             1.0        Caribbean Restaurant   
27         North York             1.0          Athletics & Sports   
28         North York             1.0                 Coffee Shop   
29          East York             1.0              Sandwich Place   
30   Downtown Toronto             1.0                        Café   
31       West Toronto             1.0                      Bakery   
32        Scarborough             1.0                  Playground   
33         North York             1.0              Clothing Store   
34         North York             1.0          Miscellaneous Shop   
..                ...             ...                         ...   
70          Etobicoke             1.0                 Pizza Place   
71        Scarborough             1.0           Accessories Store   
72         North York             1.0                 Pizza Place   
73    Central Toronto             1.0                 Coffee Shop   
74    Central Toronto             1.0                        Café   
75       West Toronto             1.0              Breakfast Spot   
76        Mississauga             1.0                       Hotel   
77          Etobicoke             1.0                 Pizza Place   
78        Scarborough             1.0                      Lounge   
79    Central Toronto             1.0                Dessert Shop   
80   Downtown Toronto             1.0                        Café   
81       West Toronto             1.0                        Café   
82        Scarborough             1.0                 Pizza Place   
83    Central Toronto             1.0                  Playground   
84   Downtown Toronto             1.0                        Café   
86    Central Toronto             1.0                 Coffee Shop   
87   Downtown Toronto             1.0              Airport Lounge   
88          Etobicoke             1.0                 Coffee Shop   
89          Etobicoke             1.0                 Pizza Place   
90        Scarborough             1.0        Fast Food Restaurant   
92   Downtown Toronto             1.0          Seafood Restaurant   
93          Etobicoke             1.0                 Pizza Place   
94          Etobicoke             1.0                   Drugstore   
96   Downtown Toronto             1.0                      Bakery   
97   Downtown Toronto             1.0                        Café   
98          Etobicoke             1.0                        Pool   
99   Downtow

Cluster 3

In [113]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
45      Department Store     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
45               Dog Run    Distribution Center

Cluster 4

In [114]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0            Print Shop           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

Cluster 5

In [115]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

We have successfully cluster Toronto neighbourhood based on venue categories!